src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -1989,17 +1989,19 @@ static ssize_t
 read_comm(ssize_t (*next_read)(int fd, void *buf, size_t size), int fd, void *buf, size_t size)
 {
 	ssize_t rc;
+	int	fd_directed;
 
 	if (!hook_enabled)
 		return next_read(fd, buf, size);
 
-	if (fd >= FD_FILE_BASE) {
-		rc = pread(fd, buf, size, file_list[fd - FD_FILE_BASE]->offset);
+	fd_directed = Get_Fd_Redirected(fd);
```

Comment:
What's the reason for this?  Is dd calling read on a fd that it got as a result of dup()?

Reply:
You are right. the fd is the result of dup(), so we need to get the redirected fd which was allocated by pil4dfs.

Reply:
Looking at the code it looks like ioctl probably needs the same fix.  Can you check the rest of the functions that take a fd please.

Reply:
Right. I created a ticket for other functions, https://daosio.atlassian.net/browse/DAOS-13927
Thank you!

Reply:
This ticket is not a priority for me, I'd rather see a single PR to address all the issues.

Reply:
ok. I will put all related changes in this PR then.